specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,143 @@
+
+The **guaranteed gas** is composed of a gas stipend, and of any guaranteed gas the user
+would like to purchase (on L1) on top of that.
+
+## Gas Stipend
+
+Because there is some cost to submitting the transaction and updating the basefee,
+we provide transactions with a small amount of free gas.
+
+If the user requests more `guaranteedGas` than the `gasStipend`, that gas will
+be bought with L1 ETH via a gas burn or by buying it directly. If they request
+less gas than the stipend, they will not be charged.
+
+We provide XX,XXX amount of gas in the stipend. TODO - should this be price
+rather than gas based??
+
+## Limiting Guaranteed Gas
+
+The total amount of guaranteed gas that can be bought in a single L1 block must
+be limited to prevent a denial of service attack against L2 as well as allow the
+total amount of guaranteed gas to be below the L2 block gas limit.
+
+We set limit the total amount of gas buyable via a contract method. It will initially
+be controlled by the Optimism Multisig before being handed over to governance.
+TODO - check that this is the actual plan.
+
+## 1559 Fee Market
+
+To reduce [Priority Gas Auctions](./glossary.md#priority-gas-auction) and accurately price gas, we implement a 1559
+style fee market on L1 with the following pseudocode. We also use this opporunity to
+place a hard limit on the amount of guaranteed gas that is provided.
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+GUARANTEED_GAS_LIMIT: u64, SANITY_GAS_LIMIT: u64 = load_and_unpack_storage2()
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
+    if curr_bought_gas == gas_target {
+        new_basefee := curr_basefee
+    } else if curr_bought_gas > gas_target {
+        gas_delta     := curr_bought_gas - gas_target
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        basefee_delta := max(basefee_delta, 1) # TODO: Why does 1559 have this asymmetry?
+        new_basefee   := curr_basefee + basefee_delta
+    } else {
+        gas_delta     := gas_target - curr_bought_gas
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+    # Fun fact, geth doesn't let the new_basefee get below 0 and while not in the EIP spec, we should add this as well.
+        new_basefee   := curr_basefee - basefee_delta
+    }
+    curr_basefee := new_basefee
+    curr_number := block.number
+    curr_bought_gas := 0
+   
+}
+
+curr_bought_gas += required_gas
+require(curr_bought_gas <= min(GUARANTEED_GAS_LIMIT, SANITY_GAS_LIMIT)
+gas_cost = requested_gas * curr_basefee
```

Comment:
`gas_cost` is missing `mint`, is this required?

Reply:
I don't think I understand the question. If you mean `gas_cost` + `mint` >= `msg.value`, note that we set `mint` to be `msg.value`.